Add addressable_light display platform

justfalter opened the pull request:
## Description:

Adds `addressable_light` as a supported `display` platform. This allows any component derived from `AddressableLight` (ex: `fastled_clockless`, `neopixelbus`, etc) to be used as a display.  I wrote the code initially for a 64-led WS2812B strip arranged in an 8x8 matrix. My next project is to build a lametric-inspired clock/weather/notification display using an 8x32 WS2812B matrix.

Notes:
- `addressable_light` requires an `AddressableLight`-derived light be defined, and supplied via the `addressable_light_id` property. 
- The display may be enabled or disabled via `set_enabled(bool)` -- allowing one to switch control back and forth between the Display mode and that of the light (maybe you want to be able to flip between lighting effects and displaying images/text/etc).

Here's an example of my 8x8 using some of the `display` component functionality to draw a bulls-eye pattern:
```
esphome:
  name: ledbox01
  platform: ESP32
  board: esp32doit-devkit-v1
  on_boot:
    priority: -10
    then:
      - switch.turn_on: control_display

light:
  - platform: fastled_clockless
    chipset: WS2812B
    pin: GPIO4
    num_leds: 64
    rgb_order: GRB
    name: "led_matrix"
    id: led_matrix
    default_transition_length: 0s
    color_correct: [50%, 50%, 50%]
    internal: false
    restore_mode: ALWAYS_ON

display:
  - platform: addressable_light
    id: my_display
    addressable_light_id: led_matrix
    width: 8
    height: 8
    rotation: 180°
    update_interval: 16ms
    lambda: |-
      Color red = Color(0xFF0000);
      Color green = Color(0x00FF00);
      Color blue = Color(0x0000FF);
      it.rectangle(0, 0, 8, 8, red);
      it.rectangle(1, 1, 6, 6, green);
      it.rectangle(2, 2, 4, 4, blue);
      it.rectangle(3, 3, 2, 2, red);

switch:
  - platform: template
    id: control_display
    name: "Control Display"
    turn_on_action:
      - lambda: |-
          id(my_display).set_enabled(true);
    turn_off_action:
      - lambda: |-
          id(my_display).set_enabled(false);
```

![bullseye](https://user-images.githubusercontent.com/120545/92350862-d8716d80-f09f-11ea-8496-79f807ee5e51.png)

**Related issue (if applicable):** 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#755

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

justfalter committed:
Add addressable_light_display

- Adds addressable_light_display, which allows any addressable_light
component to be made available as a display.

justfalter commented:
I'll get documentation written up and submitted, tomorrow.

justfalter committed:
Address clang-tidy concerns

justfalter committed:
Address python lint issues

justfalter committed:
addressable_light_display: effect handling

- Disable light effects when enabling.
- Restore last configured effect when disabling.

justfalter committed:
Address linting, again.

justfalter committed:
Fix tidy issues

justfalter committed:
Add tests for addressable_light_display

justfalter commented:
Hi there,
Just wondering if I'm missing anything that might hold up this PR? I don't know what the timeline for merging/review typically is with esphome, but any feedback would be greatly appreciated!

jesserockz commented:
@justfalter Sorry nobody has replied on this PR. I myself am waiting for an 8x32 led matrix to arrive and will be testing this out when it does. Will update you once I have it running (or having issues).

jesserockz committed:
Merge branch 'dev' into addressable_light_display

docteurzoidberg commented:
will try it on my 25x35 matrix and tell!

docteurzoidberg commented:
test on a 8x32 matrix. with untouched width and height : https://i.imgur.com/Oyxxx1K.png
test on a 8x32 matrix, when setting 8 and 32 in height/width: https://i.imgur.com/YDO1JPd.png

justfalter commented:
Since creating this PR, I had picked up one of those 8x32 ws2812b matrixes. I've got some additional changes that I need to push up that allow you to specify a custom pixel-mapper lambda from the yaml configuration.

justfalter committed:
Add pixel_mapper to addressable_light_display

- Adds support for a custom pixel mapping algorithm via `pixel_mapper`.
  The lambda receives the x and y offset, and is expected to return the
  offset into the addressable light strip.

justfalter committed:
Fix indent

justfalter commented:
I still have to update the documentation, but I hope that diff around the `pixel_mapper` makes it apparent what is going on for testing purposes 

The 8x32 ws2812b matrix that I have kind of snakes down, up, down .... starting from the upper left.
```
    X
    0  1  2  3  4
Y 0 *  |--|  |--|
  1 |  |  |  |  |
  2 |  |  |  |  |
  3 |  |  |  |  |
  4 |  |  |  |  |
  5 |  |  |  |  |
  6 |  |  |  |  |
  7 |--|  |--|  |--(etc)
```
In my case, the following worked for pixel mapping my matrix:
```
    pixel_mapper: |-
      if (x % 2 == 0) {
        return (x * 8) + y;
      }
      return (x * 8) + (7 - y);
```

docteurzoidberg commented:
perfect. will test and report ! many thanks for the update

docteurzoidberg commented:
seems to work nice with the provided  pixel mapper lambda for my 8x32 matrix too! (round one sorry): https://i.imgur.com/759tZOu.png

docteurzoidberg commented:
> 
> 
> I still have to update the documentation, but I hope that diff around the `pixel_mapper` makes it apparent what is going on for testing purposes
> 
> The 8x32 ws2812b matrix that I have kind of snakes down, up, down .... starting from the upper left.
> 
> ```
>     X
>     0  1  2  3  4
> Y 0 *  |--|  |--|
>   1 |  |  |  |  |
>   2 |  |  |  |  |
>   3 |  |  |  |  |
>   4 |  |  |  |  |
>   5 |  |  |  |  |
>   6 |  |  |  |  |
>   7 |--|  |--|  |--(etc)
> ```
> 
> In my case, the following worked for pixel mapping my matrix:
> 
> ```
>     pixel_mapper: |-
>       if (x % 2 == 0) {
>         return (x * 8) + y;
>       }
>       return (x * 8) + (7 - y);
> ```

many thanks!

justfalter commented:
I've updated the [docs PR](https://github.com/esphome/esphome-docs/pull/755) to include details on the `pixel_mapper`.

Thanks for checking out my PR, I'm glad it's working for more than just myself!

jesserockz committed:
Merge branch 'dev' into addressable_light_display

jesserockz commented on esphome/components/addressable_light_display/display.py:
```diff
@@ -0,0 +1,44 @@
+    CONF_ID, CONF_LAMBDA, CONF_PAGES, CONF_ADDRESSABLE_LIGHT_ID, CONF_HEIGHT, CONF_WIDTH,
+    CONF_UPDATE_INTERVAL, CONF_PIXEL_MAPPER
+)
+
```
```suggestion
CODEOWNERS = ["@justfalter"]

```

jesserockz reviewed (changes requested):
@justfalter I am still waiting for my matrix to arrive, but all is looking good Just a few points:
- Could you add yourself as a CODEOWNER as below and run `script/build_codeowners.py`
- Also please reformat the python file using `black`

justfalter committed:
Update esphome/components/addressable_light_display/display.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

justfalter committed:
Reformat addressable_light_display/display.py

- Per code review, reformat `display.py` using black

justfalter commented:
@jesserockz Done! Thanks, again, for keeping this going!

justfalter committed:
Update CODEOWNERS

jesserockz reviewed (changes requested):
I was just thinking, I don't like the `_display` on the end. Maybe rename it to just be `addressable_light` I know it might sound confusing because of the addressable lights, but basically this is just the display platform for those, so it does make sense:

```yaml
display:
  - platform: addressable_light
     ...
```

docteurzoidberg commented:
@justfalter : I use your component a lot !thanks for your work, it's working pretty well ! I have a tons of thing to show you on my matrices :)

docteurzoidberg commented:
@justfalter @jesserockz i wrote a "DisplayEffects" concept like in leds effets but for 2d display, but since it use a lot of maths on some effects. maybe it should be restricted to addressable light displays (because small resolutions) and become AdresseLightDisplayEffects ?

justfalter committed:
Prefer int to IntRef

justfalter committed:
Remove redundant test, reduce code size

jesserockz commented:
Sorry about the test size, you could try to move to test4.yaml which has a lot less in it.

justfalter committed:
Rename to addressable_light

justfalter committed:
Move addressable_light tests to test4

jesserockz commented on esphome/components/addressable_light/addressable_light_display.cpp:
```diff
@@ -2,9 +2,9 @@
+namespace addressable_light {
 
-static const char* TAG = "display.addressable_light_display";
+static const char* TAG = "display.addressable_light";
```
Sorry, this is something that needs a little bit of tidy up in the whole repo, but it should be the other way around
```suggestion
static const char *TAG = "addressable_light.display";
```

justfalter committed:
Update esphome/components/addressable_light/addressable_light_display.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

justfalter commented:
@jesserockz All done!

justfalter committed:
Address clang lint

jesserockz commented on esphome/components/addressable_light/addressable_light_display.cpp:
```diff
@@ -4,7 +4,7 @@
 namespace addressable_light {
 
-static const char *TAG = "addressable_light.display";
+static const char* TAG = "addressable_light.display";
```
Interesting, most of the codebase has it on the right, oh well =)

justfalter replied:
You win some, you lose some.

docteurzoidberg commented:
🥇

escoand commented:
Are there some ideas to fix the flicker when turning the light output on and off? Setting `transition_length: 0s` doesn't avoid this problem completely and is also just some hacky workaround.

k-w-1 commented:
> Are there some ideas to fix the flicker when turning the light output on and off? Setting `transition_length: 0s` doesn't avoid this problem completely and is also just some hacky workaround.

For what it's worth, I've noticed on ESP8266 you need to use the RXD pin (3) (for DMA support, apparently) with ws2812s. It doesn't get rid of the transition flicker entirely, but I was getting tons of flickering on any display updates. For me, switching to pin 3 did reduce the on/off flicker (with the 0s bodge).
